test/core/iomgr/mpmcqueue_test.cc

```diff
@@ -0,0 +1,217 @@
+  for (int i = 0; i < num_threads; ++i) {
+    consumer_thds[i] = grpc_core::New<ConsumerThread>(&queue);
+    consumer_thds[i]->Start();
+  }
+
+  WorkItem** items = static_cast<WorkItem**>(
+      gpr_zalloc(THREAD_LARGE_ITERATION * sizeof(WorkItem*)));
+  for (int i = 0; i < THREAD_LARGE_ITERATION; ++i) {
+    items[i] = grpc_core::New<WorkItem>(i);
+    queue.Put(static_cast<void*>(items[i]));
+  }
+
+  gpr_log(GPR_DEBUG, "Terminating threads...");
+  for (int i = 0; i < num_threads; ++i) {
+    queue.Put(nullptr);
+  }
+  for (int i = 0; i < num_threads; ++i) {
+    consumer_thds[i]->Join();
+  }
+  gpr_log(GPR_DEBUG, "Checking and Cleaning Up...");
+  for (int i = 0; i < THREAD_LARGE_ITERATION; ++i) {
+    GPR_ASSERT(items[i]->done);
+    grpc_core::Delete(items[i]);
+  }
+  gpr_free(items);
+  for (int i = 0; i < num_threads; ++i) {
+    grpc_core::Delete(consumer_thds[i]);
+  }
+  gpr_free(consumer_thds);
+  gpr_log(GPR_DEBUG, "Done.");
+}
+
+static void test_FIFO(void) {
+  gpr_log(GPR_INFO, "test_FIFO");
+  grpc_core::InfLenFIFOQueue large_queue;
+  for (int i = 0; i < THREAD_LARGE_ITERATION; ++i) {
+    large_queue.Put(static_cast<void*>(grpc_core::New<WorkItem>(i)));
+  }
+  GPR_ASSERT(large_queue.count() == THREAD_LARGE_ITERATION);
+  for (int i = 0; i < THREAD_LARGE_ITERATION; ++i) {
+    WorkItem* item = static_cast<WorkItem*>(large_queue.Get());
+    GPR_ASSERT(i == item->index);
+    grpc_core::Delete(item);
+  }
+}
+
+static void test_many_thread(void) {
+  gpr_log(GPR_INFO, "test_many_thread");
+  const int num_work_thd = 10;
+  const int num_get_thd = 20;
+  grpc_core::InfLenFIFOQueue queue;
+  ProducerThread** work_thds = static_cast<ProducerThread**>(
+      gpr_zalloc(num_work_thd * sizeof(ProducerThread*)));
+  ConsumerThread** consumer_thds = static_cast<ConsumerThread**>(
+      gpr_zalloc(num_get_thd * sizeof(ConsumerThread*)));
+
+  gpr_log(GPR_DEBUG, "Fork ProducerThread...");
+  for (int i = 0; i < num_work_thd; ++i) {
+    work_thds[i] = grpc_core::New<ProducerThread>(
+        &queue, i * THREAD_LARGE_ITERATION, THREAD_LARGE_ITERATION);
+    work_thds[i]->Start();
+  }
+  gpr_log(GPR_DEBUG, "ProducerThread Started.");
+  gpr_log(GPR_DEBUG, "Fork Getter Thread...");
+  for (int i = 0; i < num_get_thd; ++i) {
+    consumer_thds[i] = grpc_core::New<ConsumerThread>(&queue);
+    consumer_thds[i]->Start();
+  }
+  gpr_log(GPR_DEBUG, "Getter Thread Started.");
+  gpr_log(GPR_DEBUG, "Waiting ProducerThread to finish...");
+  for (int i = 0; i < num_work_thd; ++i) {
+    work_thds[i]->Join();
+  }
+  gpr_log(GPR_DEBUG, "All ProducerThread Terminated.");
+  gpr_log(GPR_DEBUG, "Terminating Getter Thread...");
+  for (int i = 0; i < num_get_thd; ++i) {
+    queue.Put(nullptr);
+  }
+  for (int i = 0; i < num_get_thd; ++i) {
+    consumer_thds[i]->Join();
+  }
+  gpr_log(GPR_DEBUG, "All Getter Thread Terminated.");
+  gpr_log(GPR_DEBUG, "Checking WorkItems and Cleaning Up...");
```

Comment:
Where is WorkItem being checked?

Reply:
Check was done in the destructor of ProducerThread, should I mention that in a comment around this line?

Reply:
Yes, please comment it here.